doc/developer/platform/reconciliation.md

```diff
@@ -0,0 +1,59 @@
+
+Commands represent instructions the ADAPTER sends to COMPUTE instances.
+* The ADAPTER may restart independently of a COMPUTE instance.
+* Reconciling commands is the mechanism to match a COMPUTE instance's state to what the ADAPTER expects after it starting.
+* On startup, the ADAPTER provides the current configuration to the COMPUTE instances.
+* COMPUTE instances can match the configuration to their own state and only apply changes.
+
+For this purpose, we interpret the commands ADAPTER provides to COMPUTE as an append-only log.
+Upon ADAPTER restart, the new configuration is/should be a suffix of the log of all previously received commands.
+
+The log can be compacted by collapsing commands that semantically supersede previous commands.
+
+Currently, the `Command` protocol does not express enough information to provide full command reconciliation.
+* Materialize initiates tables with ephemeral information.
+  For example, `show databases` reads from a table that is unconditionally initialized with `materialize`.
+  The reconciler has no good mechanism to distinguish initialization updates from others.
+  * *Current solution:* The coordinator truncates all ephemeral tables on startup using a new instruction.
+  * We could maintain a trace of all inserts and block already-inserted data.
+    This would require us to maintain a potentially long history of inserts.
+  * We cannot clear tables by advancing their compaction frontier to `[]`.
+    This would invalidate all dataflows that depend on the table.
+  * The timestamp of inserts during initialization is the current time, not 0, so we can't use this to distinguish initialization inserts.
+* The controller assumes that dataflows are rendered with a specific initial time.
+  When restarting, we do not render previously rendered dataflows, which might violate the implicit contract expected by the coordinator.
+  The reconciler tries to bring the coordinator up-to-date by replying with a correcting `FrontiersUppers` or `TimestampBindings` response.
+  There might be a brief moment in time when the two are not synchronized and the coordinator could produce invalid commands.
+* Peeks and responses are not handled correctly.
+  When reconnecting while a peek was active, we might send a response to a coordinator that doesn't know about the peek.
+  TODO: Determine whether the coordinator should just discard replies to a peek or if the reconciler should maintain state and locally discard the responses if the initiator for a peek goes away.
+
+### State
+
+We assume that the mapping of `GlobalId` to an object is unique, i.e., after restarts they still name the same object.
+This allows us to maintain a small amount of state to enable command reconciling.
+
+* For each installed identifier, we maintain its upper frontier.
+
+## Open questions
+
+* The coordinator does not indicate the set of known IDs.
+  This makes it difficult to determine which objects to uninstall once a coordinator reconnects.
```

Comment:
In principle, if we have a partial order on `GlobalId`, we could notice as the commands *pass* an installed id, which is our indication that the id cannot now be become bound, and so can be cleaned up. We have this at the moment, but I'm not sure it is written in stone.